scipy/optimize/_bglu_dense.pyx

```diff
@@ -0,0 +1,369 @@
+
+        # form Hessenberg matrix
+        H = self.U
+        H[:, i:self.m-1] = self.U[:, i+1:self.m]  # eliminate column i
+        H[:, -1] = um  # add column corresponding with j
+
+        # convert H to upper triangular, recording elementary row operations
+        ops = np.zeros((self.m-1-i, 2))
+        self.hess_lu(H, i, ops) # hess_lu modifies ops in place
+        self.ops_list.append(ops)
+
+        self.U = H
+
+    @_consider_refactor
+    def solve(self, q, transposed = False):
+        """
+        Solve B @ v = q efficiently using factorization
+        """
+        if not self.ops_list:
+            # before any updates, solve according to Equation 5.2
+            v = lu_solve(self.plu, q, trans=transposed)
+        else:
+            if not transposed:
+                q = q[self.pi]  # paper skips this by making
+                                # "inessential assumption" of no permutation
+
+                # Equation 5.16
+                t = solve_triangular(self.L, q, lower = True,
+                                     check_finite=False, unit_diagonal=True)
+
+                # Equation 5.17
+                temp = t
+                for ops in self.ops_list:
+                    self.perform_ops(temp, ops) # modifies temp in place
+                w = temp
+
+                # Equation 5.18
+                # For whatever reason, faster to use U.T and set trans=True
+                v = solve_triangular(self.U.T, w, lower=True,
+                                     trans=True, check_finite=False)
+
+            else: # do everything transposed and in reverse order
+                t = solve_triangular(self.U.T, q, lower=True,
+                                     trans=False, check_finite=False)
+                temp = t
+                for ops in reversed(self.ops_list):
+                    self.perform_ops(temp, ops, rev = True) # mod in place
+                w = temp
+                v = solve_triangular(self.L, w, lower = True, trans=True,
+                                     check_finite=False, unit_diagonal=True)
+                v = v[self.pit]
+
+        return v
+
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    cdef void swap_rows(self, nparray H, int i):
+        """
+        Swaps row i of H with next row; represents matrix product by PI_i
+        matrix described after matrix 5.10
+        """
+        if i == 0:
+            H[i:i+2] = H[i+1::-1]
+        else:
+            H[i:i+2] = H[i+1:i-1:-1]
+
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    @cython.cdivision(True) # not really important
+    cdef double row_subtract(self, double[:, ::1] H, int i):
+        """
+        Zeros first nonzero element of row i+1 of H by subtracting appropriate
+        multiple of row i; represents matrix product by matrix 5.10. Returns
+        factor g for storage.
+        """
+        cdef double g = H[i+1, i]/H[i,i]
+#        H[i+1, i:] -= g*H[i, i:]
```

Comment:
This should be faster if we use `daxpy` from BLAS (assuming the data is always real)

Reply:
It is always real. Not noticing any speedup adding at the top:
```
cdef extern from 'cblas.h':
void daxpy 'cblas_daxpy'(int N, double A,
                         double* X, int incX,
                         double* Y, int incY) nogil
```
then calling here:
```
daxpy(H.shape[1]-i, -g, &H[i, i], 1, &H[i+1, i], 1)
```
No yellow on any of these lines.

Suggestions for improving that? Which code do you think I should keep?

Reply:
What I had in mind was using our own wrappers by going about it with 

    from scipy.linalg.cython_blas cimport daxpy

But my comment was more towards using internal LAPACK LU factorization `dgetrf` and solvers `dgetrs` or `dtrtrs` such that you don't loose time in going back and forth for the final result. 

I feel like this is some sort of an eta factorization but I didn't read all of it so I don't know if we can directly use any existing LAPACK routines such as this one here which is essentially what you are doing with the hessenberg to triangular reduction (well emphasis on essential :) ).

https://github.com/scipy/scipy/blob/4fcd7f43546a70ee1e753dc13e5e7be4cdf4dae8/scipy/optimize/_lsq/givens_elimination.pyx

Reply:
@ilayn Thanks; didn't know about those wrappers and it was not the first hit when I was searching for using BLAS routines directly from Cython : ) I replaced my wrappers for `daxpy` and `dswap` with those from SciPy as suggested. No speed impact, although hopefully they will build properly on the cloud. 

Do you really think it's worth the time to eliminate the Python interaction in `solve` and `update` (including using `dgetrf`, etc...)? It looks like almost all the time is spent in `hess_lu` and `perform_ops`, inside of which there is no Python interaction left. These functions are only called once or twice per simplex iteration, so unless the overhead of calling a `cdef`ed function and turning an ndarray into a memoryview is outrageous I don't see how this will help much.

Now, if there is some LAPACK functionality that can mostly _replace_ `hess_lu` and `perform_ops`, I'm all for it. You're right; it looks like some Givens rotation stuff might do the trick, as this sounds like exactly what I need:
https://scicomp.stackexchange.com/questions/23984/solving-linear-system-ax-b-with-hessenberg-matrix-using-lapack
I'll look into it.

---

English description of what's going on: I need to solve linear systems involving the basis matrix B at each iteration of the simplex algorithm. To start, I perform an LU factorization. At each iteration, one column of the basis matrix is removed, and one new column is added. Rather than replacing the exiting column with the incoming column in-place (I think this would be like the eta matrix thing?), everything to the right of the column being removed is shuffled one column left, and the new column is added at the far right. It turns out that to update the factorization, the same sort of thing is done to U: everything to the right of the column being removed is shifted to the left by one, and a new column comes in at the far right. (L is unchanged.) Shifting everything over by one column like that turns U into a Hessenberg matrix. But since we're going to do this update multiple times (and since I'm unaware of an efficient Hessenberg solve in LAPACK), we don't leave U Hessenberg; we factor it into a sequence of matrices that perform elementary row operations that eliminate the off diagonal, leaving behind a new upper triangular matrix. `hess_lu` does that, and it records the elementary row operations performed to transform the Hessenberg matrix into the new upper triangular matrix. Then `perform_ops` does the same row operations on the right hand side when the time has come to solve the system.

Reply:
@ilayn I think GIvens rotations would work but I don't think there is an advantage. To reduce the Hessenberg matrix to upper triangular it looks like we'd be calling the routines for generating and applying GIvens rotations O(m) times instead of calling `dswap`/`daxpy` O(m) times (and putting some coefficients in an array). I don't imagine the Givens rotations routines could be much faster than `dswap` and `daxpy`?